File: pcc/py_runtime/src/py_tuple.c
/* pcc/py_runtime/src/py_tuple.c
 *
 * Fixed-length immutable sequence of PyObject*.
 *
 * Phase 2 layout:
 *
 *   typedef struct {
 *       PyObjectHeader h;
 *       int64_t  len;
 *       PyObject *items[];   // flexible array of owned refs
 *   } PyTupleObject;
 *
 * The header and payload live in one contiguous allocation, which matches
 * the contract exactly (see docs/plans/python-frontend-interfaces.md §3).
 * Deallocation is centralized in py_obj.c's dispatch table, so all we do
 * here is allocate, set (during construction), get (with INCREF) and
 * return len.
 *
 * Reference-counting rules:
 *   - py_tuple_new(n) zeros items[]; caller populates via set_item.
 *   - py_tuple_set_item INCREFs the stored item. The contract says "does
 *     NOT INCREF because tuple owns" in the sense that the caller is NOT
 *     donating its own ref — we take a fresh one so the caller still owns
 *     theirs. This matches Phase-1 behavior; simpler than a steal-protocol
 *     and makes construction from borrowed refs safe.
 *   - py_tuple_get returns a fresh INCREF'd ref that the caller owns.
 */

#include "py_internal.h"
#include <stdlib.h>
#include <string.h>

static int tuple_is_none_or_null(PyObject *o) {
    return o == NULL || o == py_None;
}

static int py_tuple_item_can_participate_in_cycle_with_depth(PyObject *o, int depth) {
    if (o == NULL || PY_IS_TAGGED_INT(o)) return 0;
    if (depth < 0) depth = 0;
    if (depth > 64) return 0;

    int32_t tag = py_header(o)->type_tag;
    if (tag >= PY_TYPE_USER) return 1;
    switch (tag) {
        case PY_TYPE_LIST:
        case PY_TYPE_DICT:
        case PY_TYPE_SET:
        case PY_TYPE_FUNC:
        case PY_TYPE_CLASS:
        case PY_TYPE_INSTANCE:
        case PY_TYPE_EXC:
        case PY_TYPE_ITER:
        case PY_TYPE_GEN:
        case PY_TYPE_COROUTINE:
        case PY_TYPE_MEMORYVIEW:
        case PY_TYPE_WEAKREF:
            return 1;
        case PY_TYPE_TUPLE: {
            PyTupleObject *t = (PyTupleObject *)o;
            for (int64_t i = 0; i < t->len; i++) {
                PyObject *child = pcc_gc_load_ptr(o, &t->items[i]);
                if (py_tuple_item_can_participate_in_cycle_with_depth(child, depth + 1)) {
                    return 1;
                }
            }
            return 0;
        }
        default:
            return 0;
    }
}

static int py_tuple_item_can_participate_in_cycle(PyObject *o) {
    return py_tuple_item_can_participate_in_cycle_with_depth(o, 0);
}

PyObject *py_tuple_new(int64_t n) {
    if (n < 0) n = 0;
    /* One contiguous allocation: header + n * sizeof(PyObject*). */
    size_t bytes = sizeof(PyTupleObject) + (size_t)n * sizeof(PyObject *);
    PyTupleObject *t = (PyTupleObject *)pcc_gc_alloc(
        (int64_t)bytes, PY_TYPE_TUPLE, 0
    );
    if (t == NULL) return NULL;
    t->len        = n;
    if (n > 0) {
        memset(t->items, 0, (size_t)n * sizeof(PyObject *));
    }
    return (PyObject *)t;
}

void py_tuple_set_item(PyObject *tuple, int64_t i, PyObject *item) {
    /* Used during construction only. The slot is expected to be NULL —
     * we don't decref the old slot, only incref the new item so the
     * caller's ref is independent of the tuple's. */
    if (tuple == NULL) return;
    PyTupleObject *t = (PyTupleObject *)tuple;
    if (i < 0 || i >= t->len) return;
    pcc_gc_store_ptr(tuple, &t->items[i], item);
    if (py_tuple_item_can_participate_in_cycle(item)) {
        int32_t flags = py_header(tuple)->flags;
        if ((flags & PY_FLAG_GC_TRACKED) == 0) {
            py_gc_track(tuple);
        }
    }
}

PyObject *py_tuple_get(PyObject *tuple, int64_t i) {
    if (tuple == NULL) return NULL;
    PyTupleObject *t = (PyTupleObject *)tuple;
    /* Python allows negative indices. */
    if (i < 0) i += t->len;
    if (i < 0 || i >= t->len) {
        /* TODO(phase3): raise IndexError. */
        return NULL;
    }
    PyObject *v = pcc_gc_load_ptr(tuple, &t->items[i]);
    py_incref(v);
    return v;
}

int64_t py_tuple_len(PyObject *tuple) {
    if (tuple == NULL) return 0;
    return ((PyTupleObject *)tuple)->len;
}

static int64_t tuple_slice_count(int64_t lo, int64_t hi, int64_t step) {
    int64_t count = 0;
    if (step > 0) {
        for (int64_t i = lo; i < hi; i += step) count++;
    } else {
        for (int64_t i = lo; i > hi; i += step) count++;
    }
    return count;
}

PyObject *py_tuple_slice(PyObject *tuple, PyObject *lo, PyObject *hi, PyObject *step) {
    if (tuple == NULL) return NULL;
    PyTupleObject *t = (PyTupleObject *)tuple;
    int64_t len = t->len;

    int64_t step_v = 1;
    if (!tuple_is_none_or_null(step)) {
        step_v = py_obj_index_i64(step);
        if (py_err_occurred()) return NULL;
        if (step_v == 0) {
            return NULL;
        }
    }

    int64_t lo_v, hi_v;
    if (step_v > 0) {
        lo_v = tuple_is_none_or_null(lo) ? 0   : py_obj_index_i64(lo);
        if (py_err_occurred()) return NULL;
        hi_v = tuple_is_none_or_null(hi) ? len : py_obj_index_i64(hi);
        if (py_err_occurred()) return NULL;
    } else {
        lo_v = tuple_is_none_or_null(lo) ? len - 1 : py_obj_index_i64(lo);
        if (py_err_occurred()) return NULL;
        hi_v = tuple_is_none_or_null(hi) ? -1      : py_obj_index_i64(hi);
        if (py_err_occurred()) return NULL;
    }

    if (step_v > 0) {
        if (lo_v < 0) {
            lo_v += len;
            if (lo_v < 0) lo_v = 0;
        }
        if (lo_v > len) lo_v = len;
        if (hi_v < 0) {
            hi_v += len;
            if (hi_v < 0) hi_v = 0;
        }
        if (hi_v > len) hi_v = len;
    } else {
        if (lo_v < 0) {
            lo_v += len;
            if (lo_v < 0) lo_v = -1;
        }
        if (lo_v >= len) lo_v = len - 1;

        if (hi_v < 0) {
            if (tuple_is_none_or_null(hi)) {
                hi_v = -1;
            } else {
                hi_v += len;
                if (hi_v < 0) hi_v = -1;
            }
        }
        if (hi_v >= len) hi_v = len - 1;
    }

    int64_t count = tuple_slice_count(lo_v, hi_v, step_v);
    PyObject *out = py_tuple_new(count);
    if (out == NULL) return NULL;

    int64_t j = 0;
    if (step_v > 0) {
        for (int64_t i = lo_v; i < hi_v; i += step_v) {
            PyObject *v = pcc_gc_load_ptr(tuple, &t->items[i]);
            py_tuple_set_item(out, j++, v);
        }
    } else {
        for (int64_t i = lo_v; i > hi_v; i += step_v) {
            if (i < 0 || i >= len) break;
            PyObject *v = pcc_gc_load_ptr(tuple, &t->items[i]);
            py_tuple_set_item(out, j++, v);
        }
    }
    return out;
}

PyObject *py_tuple_concat(PyObject *a, PyObject *b) {
    if (a == NULL || b == NULL) return NULL;
    PyTupleObject *ta = (PyTupleObject *)a;
    PyTupleObject *tb = (PyTupleObject *)b;
    int64_t na = ta->len;
    int64_t nb = tb->len;
    PyObject *out = py_tuple_new(na + nb);
    if (out == NULL) return NULL;
    for (int64_t i = 0; i < na; i++) {
        PyObject *v = pcc_gc_load_ptr(a, &ta->items[i]);
        py_tuple_set_item(out, i, v);
    }
    for (int64_t j = 0; j < nb; j++) {
        PyObject *v = pcc_gc_load_ptr(b, &tb->items[j]);
        py_tuple_set_item(out, na + j, v);
    }
    return out;
}

PyObject *py_tuple_repeat(PyObject *tuple, int64_t count) {
    if (tuple == NULL) return NULL;
    PyTupleObject *t = (PyTupleObject *)tuple;
    int64_t n = t->len;
    int64_t repeats = count > 0 ? count : 0;
    PyObject *out = py_tuple_new(n * repeats);
    if (out == NULL) return NULL;
    int64_t dst = 0;
    for (int64_t k = 0; k < repeats; k++) {
        for (int64_t i = 0; i < n; i++) {
            PyObject *v = pcc_gc_load_ptr(tuple, &t->items[i]);
            py_tuple_set_item(out, dst++, v);
        }
    }
    return out;
}

File: pcc/py_runtime/py/py_tuple.py
"""Phase 4c.1: pcc-Python replacement for py_runtime/src/py_tuple.c.

Each public entry point is decorated with ``@c_abi_export(<name>)`` so
pcc emits the unmangled C-ABI symbol name directly. This lets the
.o produced from this module drop-in replace py_tuple.o at link time.

Layout mirrors py_tuple.c exactly:

    typedef struct {
        PyObjectHeader h;          // 16 bytes (i64 refcount, i32 tag, i32 flags)
        int64_t        len;        // 8 bytes
        PyObject      *items[];    // flexible array of owned refs
    } PyTupleObject;

Offsets: 0 refcount, 8 type_tag, 12 flags, 16 len, 24 items[0...]

Layout constants are imported from the generated ABI module.  Closed-world
native constant exports keep these as compile-time values in library objects;
there is no module-init dependency.
"""

__pcc_runtime_port__ = True

from pcc.extern import extern, c_abi_export, c_ptr, c_int32, c_int64, c_void
from pcc.py_runtime.py.py_abi_constants import (
    PYLISTOBJECT_ITEMS_OFFSET,
    PYOBJECTHEADER_FLAGS_OFFSET,
    PYOBJECTHEADER_TYPE_TAG_OFFSET,
    PYTUPLEOBJECT_ITEMS_OFFSET,
    PYTUPLEOBJECT_LEN_OFFSET,
    PYTUPLEOBJECT_SIZE,
    PY_FLAG_GC_TRACKED,
    PY_TYPE_CLASS,
    PY_TYPE_COROUTINE,
    PY_TYPE_DICT,
    PY_TYPE_EXC,
    PY_TYPE_FUNC,
    PY_TYPE_GEN,
    PY_TYPE_INSTANCE,
    PY_TYPE_ITER,
    PY_TYPE_LIST,
    PY_TYPE_MEMORYVIEW,
    PY_TYPE_SET,
    PY_TYPE_TUPLE,
    PY_TYPE_USER,
    PY_TYPE_WEAKREF,
)
from pcc.unsafe import (
    cstr,
    global_addr,
    is_tagged_int,
    load_i32,
    load_i64,
    load_ptr,
    malloc,
    memset,
    null,
    ptr_add,
    ptr_is_null,
    ptr_to_int,
    store_i32,
    store_i64,
    store_ptr,
    stack_alloc,
    untag_int,
)


py_incref         = extern("py_incref",        (c_ptr,),                             c_void)
py_decref         = extern("py_decref",        (c_ptr,),                             c_void)
py_list_len       = extern("py_list_len",      (c_ptr,),                             c_int64)
py_list_get       = extern("py_list_get",      (c_ptr, c_int64),                     c_ptr)
py_obj_len        = extern("py_obj_len",       (c_ptr,),                             c_int64)
py_obj_getitem_i64 = extern("py_obj_getitem_i64", (c_ptr, c_int64),                  c_ptr)
py_obj_eq         = extern("py_obj_eq",        (c_ptr, c_ptr),                       c_int64)
py_int_value_i64  = extern("py_int_value_i64", (c_ptr,),                             c_int64)
py_int_to_i64     = extern("py_int_to_i64",    (c_ptr, c_ptr),                      c_int64)
py_err_occurred   = extern("py_err_occurred",  (),                                   c_int64)
py_gc_track       = extern("py_gc_track",       (c_ptr,),                             c_void)
py_exc_new        = extern("py_exc_new",        (c_int64, c_ptr),                     c_ptr)
py_raise          = extern("py_raise",          (c_ptr,),                             c_void)
# py_raise increfs; a caller that created the exception must release it.
py_raise_owned = extern("py_raise_owned", (c_ptr,), c_void)
pcc_gc_store_ptr  = extern("pcc_gc_store_ptr",  (c_ptr, c_ptr, c_ptr),                c_void)
pcc_gc_load_ptr   = extern("pcc_gc_load_ptr",   (c_ptr, c_ptr),                       c_ptr)
pcc_gc_backend = extern("pcc_gc_backend", (), c_int64)
pcc_gc_scheduler_root_register_handle = extern(
    "pcc_gc_scheduler_root_register_handle", (c_ptr,), c_ptr
)
pcc_gc_scheduler_root_unregister_handle = extern(
    "pcc_gc_scheduler_root_unregister_handle", (c_ptr,), c_void
)
pcc_gc_alloc      = extern("pcc_gc_alloc",      (c_int64, c_int32, c_int32),          c_ptr)
pcc_gc_publish_initialized = extern(
    "pcc_gc_publish_initialized", (c_ptr,), c_void
)
_pcc_debug_bad_incref = extern("pcc_debug_bad_incref", (c_ptr, c_int32), c_void)
getenv = extern("pcc_platform_getenv", (c_ptr,), c_ptr)


def _debug_bad_container(o, code: int) -> None:
    if ptr_is_null(getenv(cstr("PCC_DEBUG_RUNTIME"))) == 0:
        _pcc_debug_bad_incref(o, code)


pcc_gc_pointer_is_managed = extern(
    "pcc_gc_pointer_is_managed", (c_ptr,), c_int64
)


def _ptr_can_have_header(o) -> bool:
    return pcc_gc_pointer_is_managed(o) != 0


def _ptr_is_tuple(o) -> bool:
    if not _ptr_can_have_header(o):
        return False
    return load_i32(o, PYOBJECTHEADER_TYPE_TAG_OFFSET) == PY_TYPE_TUPLE


def _tuple_is_sane(tuple_ptr, code: int) -> bool:
    if ptr_is_null(tuple_ptr) != 0:
        return False
    if not _ptr_is_tuple(tuple_ptr):
        _debug_bad_container(tuple_ptr, code)
        return False
    length: int = load_i64(tuple_ptr, PYTUPLEOBJECT_LEN_OFFSET)
    if length < 0:
        _debug_bad_container(tuple_ptr, code)
        return False
    if length > 134217728:
        _debug_bad_container(tuple_ptr, code)
        return False
    return True


def _tuple_item_can_participate_in_cycle(item, depth: int = 0) -> int:
    if ptr_is_null(item) != 0:
        return 0
    if is_tagged_int(item) != 0:
        return 0
    if not _ptr_can_have_header(item):
        return 0
    if depth < 0:
        depth = 0
    if depth > 64:
        return 0

    tag: int = load_i32(item, PYOBJECTHEADER_TYPE_TAG_OFFSET)
    if tag >= PY_TYPE_USER:
        return 1
    if tag == PY_TYPE_LIST or tag == PY_TYPE_DICT or tag == PY_TYPE_SET or tag == PY_TYPE_FUNC:
        return 1
    if tag == PY_TYPE_CLASS or tag == PY_TYPE_INSTANCE or tag == PY_TYPE_EXC or tag == PY_TYPE_ITER or tag == PY_TYPE_GEN:
        return 1
    if tag == PY_TYPE_COROUTINE or tag == PY_TYPE_MEMORYVIEW or tag == PY_TYPE_WEAKREF:
        return 1
    if tag == PY_TYPE_TUPLE:
        length: int = load_i64(item, PYTUPLEOBJECT_LEN_OFFSET)
        i: int = 0
        while i < length:
            child = pcc_gc_load_ptr(item, ptr_add(item, PYTUPLEOBJECT_ITEMS_OFFSET + i * 8))
            if _tuple_item_can_participate_in_cycle(child, depth + 1) != 0:
                return 1
            i = i + 1
        return 0
    return 0


@c_abi_export("py_tuple_new")
def py_tuple_new(n: int):
    if n < 0:
        n = 0
    if n > 134217728:
        _debug_bad_container(null(), -130)
        return null()
    bytes_total: int = PYTUPLEOBJECT_SIZE + n * 8          # OFFSET_ITEMS + n * SIZEOF_PTR
    t = pcc_gc_alloc(bytes_total, PY_TYPE_TUPLE, 0)
    if ptr_is_null(t):
        return null()
    store_i64(t, PYTUPLEOBJECT_LEN_OFFSET, n)             # len = n
    if n > 0:
        items_ptr = ptr_add(t, PYTUPLEOBJECT_ITEMS_OFFSET)
        memset(items_ptr, 0, n * 8)
    else:
        pcc_gc_publish_initialized(t)
    return t


@c_abi_export("py_tuple_from_list")
def py_tuple_from_list(lst):
    # New tuple from a pcc list's elements. Mirrors py_tuple_from_list in
    # py_tuple.c; used by the dynamic-call lowering to normalize mixed
    # ``f(a, *rest)`` argument lists to the tuple the callable ABI requires.
    if ptr_is_null(lst):
        return null()
    n: int = py_list_len(lst)
    out = py_tuple_new(n)
    if ptr_is_null(out):
        return null()
    i: int = 0
    while i < n:
        v = py_list_get(lst, i)        # new ref
        py_tuple_set_item(out, i, v)   # increfs
        py_decref(v)
        i = i + 1
    return out


@c_abi_export("py_tuple_from_splat")
def py_tuple_from_splat(seq):
    if ptr_is_null(seq):
        return null()
    if not _ptr_can_have_header(seq):
        return null()

    tag: int = load_i32(seq, PYOBJECTHEADER_TYPE_TAG_OFFSET)
    n: int = -1
    if tag == PY_TYPE_TUPLE:
        n = load_i64(seq, PYTUPLEOBJECT_LEN_OFFSET)
    elif tag == PY_TYPE_LIST:
        n = py_list_len(seq)
    else:
        n = py_obj_len(seq)
        if py_err_occurred() != 0:
            return null()
    if n < 0:
        py_raise_owned(py_exc_new(6, cstr("tuple() argument is not iterable")))  # PY_EXC_TYPEERROR
        return null()

    out = py_tuple_new(n)
    if ptr_is_null(out):
        return null()

    i: int = 0
    if tag == PY_TYPE_TUPLE:
        while i < n:
            v = pcc_gc_load_ptr(seq, ptr_add(seq, PYTUPLEOBJECT_ITEMS_OFFSET + i * 8))
            py_tuple_set_item(out, i, v)
            i = i + 1
        return out
    if tag == PY_TYPE_LIST:
        items = load_ptr(seq, PYLISTOBJECT_ITEMS_OFFSET)
        while i < n:
            v = pcc_gc_load_ptr(seq, ptr_add(items, i * 8))
            py_tuple_set_item(out, i, v)
            i = i + 1
        return out

    while i < n:
        v = py_obj_getitem_i64(seq, i)
        if ptr_is_null(v) and py_err_occurred() != 0:
            py_decref(out)
            return null()
        py_tuple_set_item(out, i, v)
        if ptr_is_null(v) == 0:
            py_decref(v)
        i = i + 1
    return out


@c_abi_export("py_tuple_set_item")
def py_tuple_set_item(tuple_ptr, i: int, item) -> None:
    if ptr_is_null(tuple_ptr) != 0:
        return
    tuple_len: int = load_i64(tuple_ptr, PYTUPLEOBJECT_LEN_OFFSET)
    if i < 0 or i >= tuple_len:
        return
    slot_offset: int = PYTUPLEOBJECT_ITEMS_OFFSET + i * 8
    slot = ptr_add(tuple_ptr, slot_offset)
    if (
        load_i32(global_addr("pcc_gc_config_initialized"), 0) != 0
        and load_i32(global_addr("pcc_gc_backend_selected"), 0) == 0
    ):
        py_incref(item)
        store_ptr(slot, 0, item)
    else:
        pcc_gc_store_ptr(tuple_ptr, slot, item)
    if _tuple_item_can_participate_in_cycle(item) != 0:
        flags = load_i32(tuple_ptr, PYOBJECTHEADER_FLAGS_OFFSET)
        if (flags & PY_FLAG_GC_TRACKED) == 0:
            py_gc_track(tuple_ptr)
    # Only colored relocation waits for complete payload publication.
    # Keep the store/ownership/tracking work above for every collector.
    if pcc_gc_backend() != 4:
        return
    complete: int = 1
    slot_index: int = 0
    while slot_index < tuple_len:
        if ptr_is_null(pcc_gc_load_ptr(
            tuple_ptr,
            ptr_add(tuple_ptr, PYTUPLEOBJECT_ITEMS_OFFSET + slot_index * 8),
        )) != 0:
            complete = 0
            slot_index = tuple_len
        else:
            slot_index = slot_index + 1
    if complete != 0:
        pcc_gc_publish_initialized(tuple_ptr)


@c_abi_export("py_tuple_get")
def py_tuple_get(tuple_ptr, i: int):
    if ptr_is_null(tuple_ptr) != 0:
        return null()
    tuple_len: int = load_i64(tuple_ptr, PYTUPLEOBJECT_LEN_OFFSET)
    if i < 0:
        i = i + tuple_len
    if i < 0 or i >= tuple_len:
        # Non-raising: internal callers rely on the silent-NULL contract.
        # User-level t[i] subscripts route to py_tuple_getitem, which raises.
        return null()
    slot_offset: int = PYTUPLEOBJECT_ITEMS_OFFSET + i * 8
    v = pcc_gc_load_ptr(tuple_ptr, ptr_add(tuple_ptr, slot_offset))
    py_incref(v)
    return v


@c_abi_export("py_tuple_get_known")
def py_tuple_get_known(tuple_ptr, i: int):
    # Adapter-only helper: generated function-call ABI args are known tuples.
    # Keep py_tuple_get's owned-ref result but skip its defensive shape checks.
    if ptr_is_null(tuple_ptr) != 0:
        return null()
    tuple_len: int = load_i64(tuple_ptr, PYTUPLEOBJECT_LEN_OFFSET)
    if i < 0:
        i = i + tuple_len
    if i < 0 or i >= tuple_len:
        return null()
    slot_offset: int = PYTUPLEOBJECT_ITEMS_OFFSET + i * 8
    v = pcc_gc_load_ptr(tuple_ptr, ptr_add(tuple_ptr, slot_offset))
    py_incref(v)
    return v


@c_abi_export("py_tuple_getitem")
def py_tuple_getitem(tuple_ptr, i: int):
    # t[i] subscript: like py_tuple_get but raises IndexError on out-of-range so
    # try/except can catch it. Mirrors py_tuple_getitem in py_tuple.c; py_tuple_get
    # stays non-raising for other callers. Negative indices normalize.
    if ptr_is_null(tuple_ptr) != 0:
        return null()
    tuple_len: int = load_i64(tuple_ptr, PYTUPLEOBJECT_LEN_OFFSET)
    if i < 0:
        i = i + tuple_len
    if i < 0 or i >= tuple_len:
        py_raise_owned(py_exc_new(5, cstr("tuple index out of range")))  # PY_EXC_INDEXERROR
        return null()
    slot_offset: int = PYTUPLEOBJECT_ITEMS_OFFSET + i * 8
    v = pcc_gc_load_ptr(tuple_ptr, ptr_add(tuple_ptr, slot_offset))
    py_incref(v)
    return v


@c_abi_export("py_tuple_len")
def py_tuple_len(tuple_ptr) -> int:
    if ptr_is_null(tuple_ptr) != 0:
        return 0
    return load_i64(tuple_ptr, PYTUPLEOBJECT_LEN_OFFSET)


def _tuple_method_prepare_root(slot, value, backend: int):
    store_ptr(slot, 0, value)
    if (
        (backend == 3 or backend == 4)
        and ptr_is_null(value) == 0
        and is_tagged_int(value) == 0
    ):
        handle = pcc_gc_scheduler_root_register_handle(slot)
        if ptr_is_null(handle) == 0:
            store_ptr(slot, 0, pcc_gc_load_ptr(null(), slot))
        return handle
    return null()


def _tuple_method_root_failed(value, backend: int, handle) -> int:
    if backend != 3 and backend != 4:
        return 0
    if ptr_is_null(value) != 0 or is_tagged_int(value) != 0:
        return 0
    return ptr_is_null(handle)


def _tuple_method_reload_root(slot, handle):
    value = load_ptr(slot, 0)
    if ptr_is_null(handle) == 0:
        value = pcc_gc_load_ptr(null(), slot)
        store_ptr(slot, 0, value)
    return value


def _tuple_method_finish_root(handle) -> None:
    if ptr_is_null(handle) == 0:
        pcc_gc_scheduler_root_unregister_handle(handle)


def _tuple_method_scan(tuple_ptr, item, start, stop, want_first: int) -> int:
    if ptr_is_null(tuple_ptr) != 0:
        return -1 if want_first != 0 else 0
    backend: int = pcc_gc_backend()
    tuple_slot = stack_alloc(8)
    query_slot = stack_alloc(8)
    element_slot = stack_alloc(8)
    tuple_handle = _tuple_method_prepare_root(tuple_slot, tuple_ptr, backend)
    if _tuple_method_root_failed(tuple_ptr, backend, tuple_handle) != 0:
        return -1 if want_first != 0 else 0
    query_handle = _tuple_method_prepare_root(query_slot, item, backend)
    if _tuple_method_root_failed(item, backend, query_handle) != 0:
        _tuple_method_finish_root(tuple_handle)
        return -1 if want_first != 0 else 0

    tuple_ptr = _tuple_method_reload_root(tuple_slot, tuple_handle)
    length: int = py_tuple_len(tuple_ptr)
    lo: int = 0
    hi: int = length
    if ptr_is_null(start) == 0 or ptr_is_null(stop) == 0:
        overflow = stack_alloc(4)
        if ptr_is_null(start) == 0:
            store_i32(overflow, 0, 0)
            lo = py_int_to_i64(start, overflow)
            if load_i32(overflow, 0) != 0:
                lo = 0
        if ptr_is_null(stop) == 0:
            store_i32(overflow, 0, 0)
            hi = py_int_to_i64(stop, overflow)
            if load_i32(overflow, 0) != 0:
                hi = length
        if lo < 0:
            lo = lo + length
            if lo < 0:
                lo = 0
        if hi < 0:
            hi = hi + length
            if hi < 0:
                hi = 0
        if hi > length:
            hi = length

    count: int = 0
    i: int = lo
    while i < hi:
        tuple_ptr = _tuple_method_reload_root(tuple_slot, tuple_handle)
        item = _tuple_method_reload_root(query_slot, query_handle)
        element = py_tuple_get(tuple_ptr, i)
        store_ptr(element_slot, 0, element)
        element_handle = _tuple_method_prepare_root(
            element_slot, element, backend
        )
        if _tuple_method_root_failed(element, backend, element_handle) != 0:
            py_decref(element)
            break
        equal: int = py_obj_eq(element, item)
        tuple_ptr = _tuple_method_reload_root(tuple_slot, tuple_handle)
        item = _tuple_method_reload_root(query_slot, query_handle)
        element = _tuple_method_reload_root(element_slot, element_handle)
        _tuple_method_finish_root(element_handle)
        py_decref(element)
        if py_err_occurred() != 0:
            _tuple_method_finish_root(query_handle)
            _tuple_method_finish_root(tuple_handle)
            return -1 if want_first != 0 else 0
        if equal != 0:
            if want_first != 0:
                _tuple_method_finish_root(query_handle)
                _tuple_method_finish_root(tuple_handle)
                return i
            count = count + 1
        i = i + 1
    _tuple_method_finish_root(query_handle)
    _tuple_method_finish_root(tuple_handle)
    if want_first != 0:
        return -1
    return count


@c_abi_export("py_tuple_count")
def py_tuple_count(tuple_ptr, item) -> int:
    return _tuple_method_scan(tuple_ptr, item, null(), null(), 0)


@c_abi_export("py_tuple_index")
def py_tuple_index(tuple_ptr, item) -> int:
    result: int = _tuple_method_scan(tuple_ptr, item, null(), null(), 1)
    if result >= 0:
        return result
    if py_err_occurred() != 0:
        return -1
    py_raise_owned(py_exc_new(2, cstr("tuple.index(x): x not in tuple")))
    return -1


@c_abi_export("py_tuple_index_range")
def py_tuple_index_range(tuple_ptr, item, start, stop) -> int:
    result: int = _tuple_method_scan(tuple_ptr, item, start, stop, 1)
    if result >= 0:
        return result
    if py_err_occurred() != 0:
        return -1
    py_raise_owned(py_exc_new(2, cstr("tuple.index(x): x not in tuple")))
    return -1


@c_abi_export("py_tuple_concat")
def py_tuple_concat(a, b):
    if not _tuple_is_sane(a, -135):
        return null()
    if not _tuple_is_sane(b, -136):
        return null()
    la: int = load_i64(a, PYTUPLEOBJECT_LEN_OFFSET)
    lb: int = load_i64(b, PYTUPLEOBJECT_LEN_OFFSET)
    out = py_tuple_new(la + lb)
    i: int = 0
    while i < la:
        v = pcc_gc_load_ptr(a, ptr_add(a, PYTUPLEOBJECT_ITEMS_OFFSET + i * 8))
        py_tuple_set_item(out, i, v)
        i = i + 1
    j: int = 0
    while j < lb:
        v = pcc_gc_load_ptr(b, ptr_add(b, PYTUPLEOBJECT_ITEMS_OFFSET + j * 8))
        py_tuple_set_item(out, la + j, v)
        j = j + 1
    return out


@c_abi_export("py_tuple_repeat")
def py_tuple_repeat(tuple_ptr, count: int):
    if not _tuple_is_sane(tuple_ptr, -137):
        return null()
    length: int = load_i64(tuple_ptr, PYTUPLEOBJECT_LEN_OFFSET)
    repeats: int = count
    if repeats < 0:
        repeats = 0
    out = py_tuple_new(length * repeats)
    if ptr_is_null(out):
        return null()
    dst: int = 0
    k: int = 0
    while k < repeats:
        i: int = 0
        while i < length:
            v = pcc_gc_load_ptr(tuple_ptr, ptr_add(tuple_ptr, PYTUPLEOBJECT_ITEMS_OFFSET + i * 8))
            py_tuple_set_item(out, dst, v)
            dst = dst + 1
            i = i + 1
        k = k + 1
    return out
